### PR TITLE
fix: close compiler on SIGINT

### LINKED ISSUE #2918: filesystem cache corruption on SIGINT
If webpack-cli is terminated by SIGINT while writing to the filesystem cache (e.g. in watch mode during an incremental rebuild of a project with large external source maps), the filesystem cache will be corrupted and must be erased before webpack will build again.

webpack-cli should [hook `SIGINT`](https://nodejs.org/api/process.html#process_signal_events) and await a clean exit from [compiler.close](https://webpack.js.org/api/node/#compiler-instance).

Note that the child processes spawned by e.g. `fork-ts-checker-webpack-plugin` or `thread-loader` will hear SIGINT from the parent process, so they'll need to be isolated by spawning the compiler into a detached child process.

### PR DESCRIPTION
**Pull request by snitin315**
**What kind of change does this PR introduce?**
fix

**Did you add tests for your changes?**
No
**If relevant, did you update the documentation?**
No
**Summary**

Fix #2918 

**Does this PR introduce a breaking change?**
None

**Other information**
No

### CODE REVIEW

File: packages/webpack-cli/lib/webpack-cli.js [hunk 1]
```diff
@@ -2125,13 +2125,18 @@ class WebpackCLI {
...
+    });
+
     const isWatch = (compiler) =>
       compiler.compilers
         ? compiler.compilers.some((compiler) => compiler.options.watch)
         : compiler.options.watch;
 
     if (isWatch(compiler) && this.needWatchStdin(compiler)) {
       process.stdin.on("end", () => {
+        compiler.close();
```
Comment (alexander-akait) [inline packages/webpack-cli/lib/webpack-cli.js:2139]: ` compiler.close(() => { process.exit(0) });`

File: packages/webpack-cli/lib/webpack-cli.js [hunk 2]
```diff
@@ -2125,13 +2125,18 @@ class WebpackCLI {
       return;
     }
 
+    process.on("SIGINT", () => {
+      compiler.close();
```
Comment (alexander-akait) [inline packages/webpack-cli/lib/webpack-cli.js:2129]: We need test on `130` code here

**Review by alexander-akait — COMMENTED** [review]
Also let's add test on 130 code and close for serve (it is already implemented in dev server)

**Review by alexander-akait — COMMENTED** [review]
Looks good, let's add tests for `build`/`watch`/`serve`, just enable logging and check on https://github.com/webpack/webpack/blob/main/lib/cache/PackFileCacheStrategy.js#L1263 or you can implement custom plugin and use `compiler.cache.hooks.shutdown.tapPromise()` with `{ stage: 9999 }` to ensure we will run after cache was stored

**Review by evenstensberg — APPROVED** [review]

**Comment by alexander-akait** [comment]
I will add tests

**Comment by alexander-akait** [comment]
@snitin315 We need improve logic here: we need allow to graceful close and allow immutability disable, my idea - when developer run `CTRL + C` we try to `close()`, if `CTRL + C` pressed again we close immediately, also we need handle ` const signals = ["SIGINT", "SIGTERM"];` signals, for `serve` we need disable `setupExitSignals` and move this logic here https://github.com/webpack/webpack-dev-server/blob/master/lib/Server.js#L1088 (in future major release we will remove this option, because it is wrong place for this, developer should handle closing manually)

**Comment by snitin315** [comment]
Makes sense, I will update it in near future.

**Comment by alexander-akait** [comment]
@snitin315 hello, let's finish, just copy logic from dev server

**Comment by snitin315** [comment]
@alexander-akait sure 👍🏻

**Review by evenstensberg — CHANGES_REQUESTED** [review]
Could you rebase?

File: packages/serve/src/index.ts [hunk 3]
```diff
@@ -132,6 +132,40 @@ class ServeCommand {
 
         const servers = [];
 
+        const stopAllServers = () => {
+          Promise.all(
```
Comment (evenstensberg) [inline packages/serve/src/index.ts:136]: I'd love to make this a variable and passing the variable to the promise expression

**Pull request closed without merging**